lucene/expressions/src/java/org/apache/lucene/expressions/ExpressionValueSource.java

```diff
@@ -42,13 +42,17 @@
     this.expression = Objects.requireNonNull(expression);
     variables = new DoubleValuesSource[expression.variables.length];
     boolean needsScores = false;
-    for (int i = 0; i < variables.length; i++) {
-      DoubleValuesSource source = bindings.getDoubleValuesSource(expression.variables[i]);
-      if (source == null) {
-        throw new RuntimeException("Internal error. Variable (" + expression.variables[i] + ") does not exist.");
+    try {
+      for (int i = 0; i < variables.length; i++) {
+        DoubleValuesSource source = bindings.getDoubleValuesSource(expression.variables[i]);
+        if (source == null) {
+          throw new RuntimeException("Internal error. Variable (" + expression.variables[i] + ") does not exist.");
+        }
+        needsScores |= source.needsScores();
+        variables[i] = source;
       }
-      needsScores |= source.needsScores();
-      variables[i] = source;
+    } catch (StackOverflowError e) {
```

Comment:
I've reworked this so that instead of a `Supplier<DoubleValuesSource>` we store a `Function<Bindings, DoubleValuesSource>`, and supply a special `Bindings` implementation in `validate()` that checks for cycles.  Definitely much nicer, thanks for nudging me in the right direction!